### PR TITLE
[Merged by Bors] - feat(Mathlib/Topology/Bases): subbasis closed under intersection is a basis

### PR DESCRIPTION
**Pull request by mans0954**
We show that if a sub-basis is closed under finite intersections, then it is a basis for a topology.

As a corollary, if a sub-basis is closed under intersections, then inserting the universal set gives a basis for the topology.

An example application of this result is given in #12234

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Bases.lean [hunk 1]
```diff
@@ -119,6 +120,39 @@ theorem isTopologicalBasis_of_subbasis {s : Set (Set α)} (hs : t = generateFrom
...
+    (hsi : FiniteInter s) : IsTopologicalBasis s := by
+  have e1 {s : Set (Set α)} (hsi : FiniteInter s) :
+    (fun g => ⋂₀ g) '' { f : Set (Set α) | f.Finite ∧ f ⊆ s } = s := by
+    rw [le_antisymm_iff]
+    constructor
+    · intro F
+      simp only [mem_image, mem_setOf_eq, forall_exists_index, and_imp]
+      intro G hG hGs hi
+      rw [← hi]
+      let G' := Finite.toFinset hG
```
Comment (mans0954) [inline Mathlib/Topology/Bases.lean:133]: I'm sure there must be a better argument, but I haven't been able to figure it out.

File: Mathlib/Data/Set/Constructions.lean [hunk 2]
```diff
@@ -82,4 +82,23 @@ theorem finiteInterClosure_insert {A : Set α} (cond : FiniteInter S) (P)
...
+      · apply Or.inl
+        rw [hsu, htu, inter_self]
+      · apply Or.inr
+        rw [hsu, univ_inter]
+        exact htS
+    · apply Or.inr
+      rcases ht with htu | htS
+      · rw [htu, inter_univ]
+        exact hsS
+      · apply h hsS htS
```
Comment (j-loreaux) [inline Mathlib/Data/Set/Constructions.lean:102]:
```suggestion
  inter_mem s hs t ht:= by
    obtain ⟨(rfl | hs), (rfl | ht)⟩ := And.intro hs ht
    all_goals aesop
```

**Review by j-loreaux — COMMENTED** [review]
back for more in a bit.

File: Mathlib/Topology/Bases.lean [hunk 3]
```diff
@@ -119,6 +120,39 @@ theorem isTopologicalBasis_of_subbasis {s : Set (Set α)} (hs : t = generateFrom
...
+  exact isTopologicalBasis_of_subbasis hsg
+
+theorem isTopologicalBasis_of_subbasis_of_inter {r : Set (Set α)} (hsg : t = generateFrom r)
+    (hsi : ∀ ⦃s⦄, s ∈ r → ∀ ⦃t⦄, t ∈ r → s ∩ t ∈ r) : IsTopologicalBasis (insert univ r) := by
+  apply isTopologicalBasis_of_subbasis_of_finiteInter _ (FiniteInter.mk₂ hsi)
+  rw [hsg]
+  apply le_antisymm _ (generateFrom_anti <| subset_insert univ r)
+  exact le_generateFrom fun _ hs => (by
+    rcases hs with h | h; simp only [h, isOpen_univ];
+    exact isOpen_generateFrom_of_mem h)
```
Comment (j-loreaux) [inline Mathlib/Topology/Bases.lean:154]: Just extract the lemma and the `simp` lemma here (of course, put it in the appropriate place):
```lean
lemma generateFrom_insert_of_generateOpen {α : Type*} {s : Set (Set α)} {t : Set α}
    (ht : GenerateOpen s t) : generateFrom (insert t s) = generateFrom s := by
  refine le_antisymm (generateFrom_anti <| subset_insert t s) (le_generateFrom ?_)
  rintro t (rfl | h)
  · exact ht
  · exact isOpen_generateFrom_of_mem h

@[simp]
lemma generateFrom_insert_univ {α : Type*} {s : Set (Set α)} :
    generateFrom (insert univ s) = generateFrom s :=
  generateFrom_insert_of_generateOpen .univ
```
And then this proof becomes:
```suggestion
    (hsi : ∀ ⦃s⦄, s ∈ r → ∀ ⦃t⦄, t ∈ r → s ∩ t ∈ r) : IsTopologicalBasis (insert univ r) :=
  isTopologicalBasis_of_subbasis_of_finiteInter (by simpa using hsg) (FiniteInter.mk₂ hsi)
```

**Review by j-loreaux — COMMENTED** [review]
Here are some simplifications for you.

File: Mathlib/Topology/Bases.lean (hunk 3 above)
Comment (mans0954) [reply Mathlib/Topology/Bases.lean]: I assume the correct place for the lemmas is in `Topology/Order`?

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**